lib/matrix.js

```diff
@@ -52,11 +58,351 @@ var init = function(exports){
+    MatrixInMemoryStore.prototype = {
+                
+        /*
+         * Add an array of one or more state MatrixEvents into the store, overwriting
+         * any existing state with the same {room, type, stateKey} tuple.
+         */
+        setStateEvents: function(stateEvents) {
+            // we store stateEvents indexed by room, event type and state key.
+            for (var i = 0; i < stateEvents.length; i++) {
+                var event = stateEvents[i].event;
+                var roomId = event.room_id;
+                if (this.rooms[roomId] === undefined) {
+                    this.rooms[roomId] = {};
+                }
+                if (this.rooms[roomId].state === undefined) {
+                    this.rooms[roomId].state = {};
+                }
+                if (this.rooms[roomId].state[event.type] === undefined) {
+                    this.rooms[roomId].state[event.type] = {};
+                }
+                this.rooms[roomId].state[event.type][event.state_key] = stateEvents[i];
+            }
+        },
+       
+        /*
+         * Return a list of MatrixEvents from the store, filtered by roomid, type and state key.
+         * @param {String} roomId the Room ID whose state is to be returned
+         * @param {String} type the type of the state events to be returned (optional)
+         * @param {String} stateKey the stateKey of the state events to be returned
+         *                 (optional, requires type to be specified)
+         */
+        getStateEvents: function(roomId, type, stateKey) {
+            var stateEvents = [];
+            if (stateKey === undefined && type === undefined) {
+                for (type in this.rooms[roomId].state) {
+                    if (this.rooms[roomId].state.hasOwnProperty(type)) {
+                        for (stateKey in this.rooms[roomId].state[type]) {
+                            if (this.rooms[roomId].state[type].hasOwnProperty(stateKey)) {
+                                stateEvents.push(this.rooms[roomId].state[type][stateKey]);
+                            }
+                        }
+                    }
+                }                    
+                return stateEvents;
+            }
+            else if (stateKey === undefined) {
+                for (stateKey in this.rooms[roomId].state[type]) {
+                    if (this.rooms[roomId].state[type].hasOwnProperty(stateKey)) {
+                        stateEvents.push(this.rooms[roomId].state[type][stateKey]);
+                    }
+                }
+                return stateEvents;
+            }
+            else {
+                return [this.rooms[roomId].state[type][stateKey]];
+            }
+        },
+        
+        /*
+         * Return a single state MatrixEvent from the store for the given roomId
+         * and type.  stateKey is optional; if missing, is assumed to be blank.
+         */
+        getStateEvent: function(roomId, type, stateKey) {
```

Comment:
Missing jsdocs, which is particularly crucial here because of ``stateKey``. The description isn't clear what "blank" means for ``if missing, is assumed to be blank.`` Do you mean ``undefined`` or a 0-length string `""`?